File: app/clojure/test/ch10/concurrency_test.clj
(ns ch10.concurrency-test
  (:require [clojure.test :refer [deftest testing is are]]
            [ch10.concurrency :as cc]))

;; =============================================================================
;; City と CityStats のテスト
;; =============================================================================

(deftest city-tests
  (testing "city creation"
    (let [c (cc/city "Sydney")]
      (is (= "Sydney" (:name c))))))

(deftest city-stats-tests
  (testing "city-stats creation"
    (let [c (cc/city "Sydney")
          stats (cc/city-stats c 100)]
      (is (= c (:city stats)))
      (is (= 100 (:check-ins stats))))))

;; =============================================================================
;; top-cities のテスト
;; =============================================================================

(deftest top-cities-tests
  (testing "top-cities returns sorted cities"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")
          lima (cc/city "Lima")
          check-ins {sydney 100 dublin 200 lima 50}
          result (cc/top-cities check-ins 3)]
      (is (= 3 (count result)))
      (is (= dublin (:city (first result))))
      (is (= sydney (:city (second result))))
      (is (= lima (:city (nth result 2))))))

  (testing "top-cities with fewer results"
    (let [sydney (cc/city "Sydney")
          check-ins {sydney 100}
          result (cc/top-cities check-ins 3)]
      (is (= 1 (count result))))))

;; =============================================================================
;; 逐次処理のテスト
;; =============================================================================

(deftest process-check-ins-sequential-tests
  (testing "sequential processing counts correctly"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")
          check-ins [sydney sydney dublin sydney]
          result (cc/process-check-ins-sequential check-ins)]
      (is (= 3 (get result sydney)))
      (is (= 1 (get result dublin))))))

;; =============================================================================
;; 並行処理のテスト
;; =============================================================================

(deftest store-check-in-tests
  (testing "store-check-in! updates atom"
    (let [stored (atom {})
          sydney (cc/city "Sydney")]
      (cc/store-check-in! stored sydney)
      (cc/store-check-in! stored sydney)
      (is (= 2 (get @stored sydney))))))

(deftest process-check-ins-concurrent-tests
  (testing "concurrent processing counts correctly"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")
          check-ins (concat (repeat 100 sydney) (repeat 50 dublin))
          result (cc/process-check-ins-concurrent check-ins)]
      (is (= 100 (get result sydney)))
      (is (= 50 (get result dublin))))))

;; =============================================================================
;; pmap のテスト
;; =============================================================================

(deftest parallel-map-tests
  (testing "parallel-map applies function"
    (let [result (cc/parallel-map inc [1 2 3 4 5])]
      (is (= [2 3 4 5 6] result)))))

;; =============================================================================
;; サイコロのテスト
;; =============================================================================

(deftest cast-the-die-tests
  (testing "cast-the-die returns valid dice value"
    (dotimes [_ 100]
      (let [result (cc/cast-the-die)]
        (is (>= result 1))
        (is (<= result 6))))))

(deftest cast-dice-concurrent-tests
  (testing "cast-dice-concurrent returns correct count"
    (let [result (cc/cast-dice-concurrent 10)]
      (is (= 10 (count result)))
      (is (every? #(and (>= % 1) (<= % 6)) result)))))

(deftest cast-dice-and-sum-tests
  (testing "cast-dice-and-sum returns sum in valid range"
    (let [n 5
          result (cc/cast-dice-and-sum n)]
      (is (>= result n))      ; 最小: n * 1
      (is (<= result (* n 6))))))  ; 最大: n * 6

(deftest cast-dice-with-storage-tests
  (testing "cast-dice-with-storage! stores all results"
    (let [n 10
          result (cc/cast-dice-with-storage! n)]
      (is (= n (count result)))
      (is (every? #(and (>= % 1) (<= % 6)) result)))))

;; =============================================================================
;; チェックイン処理の完全版テスト
;; =============================================================================

(deftest start-check-in-processing-tests
  (testing "check-in processing starts and stops"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")
          check-ins (concat (repeat 10 sydney) (repeat 5 dublin))
          processing (cc/start-check-in-processing! check-ins)]
      ;; 少し待機
      (Thread/sleep 200)
      (let [ranking ((:current-ranking processing))]
        (is (>= (count ranking) 0)))
      ;; 停止
      ((:stop processing)))))

;; =============================================================================
;; agent のテスト
;; =============================================================================

(deftest agent-check-in-tests
  (testing "agent stores check-ins"
    (cc/agent-reset-check-ins!)
    (let [sydney (cc/city "Sydney")]
      (cc/agent-store-check-in! sydney)
      (cc/agent-store-check-in! sydney)
      (await cc/check-in-agent)
      (is (= 2 (get (cc/agent-get-check-ins) sydney))))))

;; =============================================================================
;; ref と STM のテスト
;; =============================================================================

(deftest stm-check-in-tests
  (testing "STM stores check-ins"
    (dosync (ref-set cc/check-in-ref {}))
    (let [sydney (cc/city "Sydney")]
      (cc/stm-store-check-in! sydney)
      (cc/stm-store-check-in! sydney)
      (is (= 2 (get @cc/check-in-ref sydney))))))

(deftest stm-update-ranking-tests
  (testing "STM updates ranking"
    (dosync
     (ref-set cc/check-in-ref {(cc/city "Sydney") 100
                               (cc/city "Dublin") 200})
     (ref-set cc/ranking-ref []))
    (cc/stm-update-ranking! 2)
    (let [ranking @cc/ranking-ref]
      (is (= 2 (count ranking)))
      ;; CityStats record なので :city でアクセス
      (is (= 200 (:check-ins (first ranking)))))))

(deftest stm-transfer-check-ins-tests
  (testing "STM transfers check-ins"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")]
      (dosync
       (ref-set cc/check-in-ref {sydney 100 dublin 50}))
      (is (true? (cc/stm-transfer-check-ins! sydney dublin 30)))
      (is (= 70 (get @cc/check-in-ref sydney)))
      (is (= 80 (get @cc/check-in-ref dublin)))))

  (testing "STM transfer fails if insufficient"
    (let [sydney (cc/city "Sydney")
          dublin (cc/city "Dublin")]
      (dosync
       (ref-set cc/check-in-ref {sydney 10 dublin 50}))
      (is (nil? (cc/stm-transfer-check-ins! sydney dublin 30)))
      (is (= 10 (get @cc/check-in-ref sydney))))))

;; =============================================================================
;; 並列 API 呼び出しのテスト
;; =============================================================================

(deftest parallel-api-calls-tests
  (testing "parallel API calls return results"
    (let [results (cc/parallel-api-calls ["a" "b" "c"])]
      (is (= 3 (count results)))
      (is (every? :result results)))))

(deftest parallel-api-calls-with-timeout-tests
  (testing "parallel API calls with timeout"
    (let [results (cc/parallel-api-calls-with-timeout ["a" "b"] 500)]
      (is (= 2 (count results)))
      (is (every? #(or (:result %) (:error %)) results)))))

;; =============================================================================
;; 並行カウントのテスト
;; =============================================================================

(deftest count-evens-tests
  (testing "count-evens counts correctly"
    (let [numbers [1 2 3 4 5 6 7 8 9 10]
          result (cc/count-evens numbers)]
      (is (= 5 result)))))

(deftest parallel-filter-count-tests
  (testing "parallel-filter-count counts correctly"
    (let [numbers (range 1 101)
          result (cc/parallel-filter-count #(zero? (mod % 3)) numbers)]
      (is (= 33 result)))))

;; =============================================================================
;; タイムアウト付き収集のテスト
;; =============================================================================

(deftest collect-for-tests
  (testing "collect-for collects values"
    (let [counter (atom 0)
          producer-fn #(swap! counter inc)
          result (cc/collect-for 200 producer-fn 50)]
      (is (>= (count result) 2))
      (is (<= (count result) 6)))))

;; =============================================================================
;; 並行マップ更新のテスト
;; =============================================================================

(deftest apply-updates-tests
  (testing "apply-updates applies all updates"
    (let [updates [(cc/->Update "a" 1)
                   (cc/->Update "b" 2)
                   (cc/->Update "c" 3)]
          result (cc/apply-updates updates)]
      (is (= 1 (get result "a")))
      (is (= 2 (get result "b")))
      (is (= 3 (get result "c"))))))

;; =============================================================================
;; イベントプロセッサのテスト
;; =============================================================================

(deftest event-processor-tests
  (testing "event processor handles events"
    (let [processor (cc/event-processor)
          handled (atom [])]
      ((:add-handler processor) :test #(swap! handled conj %))
      ((:emit processor) (cc/create-event :test {:value 1}))
      ((:emit processor) (cc/create-event :test {:value 2}))
      (Thread/sleep 50)
      (is (= 2 (count @handled)))
      ((:stop processor)))))

;; =============================================================================
;; 並行キャッシュのテスト
;; =============================================================================

(deftest create-cache-tests
  (testing "cache stores and retrieves values"
    (let [cache (cc/create-cache)
          compute-count (atom 0)]
      ;; 最初の取得（キャッシュミス）
      (let [result ((:get cache) :key1 #(do (swap! compute-count inc) 42))]
        (is (= 42 result)))
      ;; 2回目の取得（キャッシュヒット）
      (let [result ((:get cache) :key1 #(do (swap! compute-count inc) 100))]
        (is (= 42 result)))
      ;; compute は1回だけ呼ばれる
      (is (= 1 @compute-count))
      ;; stats を確認
      (let [stats ((:stats cache))]
        (is (= 1 (:size stats)))
        (is (= 1 (:hits stats)))
        (is (= 1 (:misses stats))))))

  (testing "cache invalidation"
    (let [cache (cc/create-cache)]
      ((:get cache) :key1 (constantly 42))
      ((:invalidate cache) :key1)
      (let [result ((:get cache) :key1 (constantly 100))]
        (is (= 100 result))))))

File: app/elixir/test/ch10/concurrency_test.exs
defmodule Ch10.ConcurrencyTest do
  use ExUnit.Case, async: true
  doctest Ch10.Concurrency

  alias Ch10.Concurrency
  alias Ch10.Concurrency.{City, CityStats, Update}

  describe "top_cities/2" do
    test "トップ N 都市を取得" do
      city_check_ins = %{
        %City{name: "Tokyo"} => 100,
        %City{name: "Sydney"} => 80,
        %City{name: "Dublin"} => 90,
        %City{name: "Lima"} => 70
      }

      top = Concurrency.top_cities(city_check_ins, 2)
      assert length(top) == 2
      assert hd(top).city.name == "Tokyo"
      assert hd(top).check_ins == 100
    end

    test "空のマップ" do
      assert Concurrency.top_cities(%{}, 3) == []
    end
  end

  describe "add_check_in/2" do
    test "新しい都市を追加" do
      city = %City{name: "Tokyo"}
      map = Concurrency.add_check_in(%{}, city)
      assert map[city] == 1
    end

    test "既存の都市をインクリメント" do
      city = %City{name: "Tokyo"}
      map = %{city => 5}
      map = Concurrency.add_check_in(map, city)
      assert map[city] == 6
    end
  end

  describe "Agent による共有状態" do
    test "チェックインストアの基本操作" do
      {:ok, agent} = Concurrency.create_check_in_store()
      city = %City{name: "Tokyo"}

      assert Concurrency.get_check_ins(agent) == %{}

      Concurrency.store_check_in(agent, city)
      assert Concurrency.get_check_ins(agent)[city] == 1

      Concurrency.store_check_in(agent, city)
      assert Concurrency.get_check_ins(agent)[city] == 2

      Agent.stop(agent)
    end

    test "ランキングストアの基本操作" do
      {:ok, agent} = Concurrency.create_ranking_store()

      assert Concurrency.get_ranking(agent) == []

      stats = [%CityStats{city: %City{name: "Tokyo"}, check_ins: 100}]
      Concurrency.update_ranking(agent, stats)

      assert Concurrency.get_ranking(agent) == stats

      Agent.stop(agent)
    end
  end

  describe "parallel_run/1" do
    test "複数のタスクを並列実行" do
      tasks = [fn -> 1 end, fn -> 2 end, fn -> 3 end]
      results = Concurrency.parallel_run(tasks)
      assert results == [1, 2, 3]
    end

    test "空リスト" do
      assert Concurrency.parallel_run([]) == []
    end
  end

  describe "parallel_map/2" do
    test "並列でマッピング" do
      results = Concurrency.parallel_map([1, 2, 3], &(&1 * 2))
      assert results == [2, 4, 6]
    end
  end

  describe "sum_parallel/1" do
    test "並列で合計" do
      ios = [fn -> 1 end, fn -> 2 end, fn -> 3 end]
      assert Concurrency.sum_parallel(ios) == 6
    end
  end

  describe "サイコロ" do
    test "cast_the_die は 1-6 を返す" do
      for _ <- 1..100 do
        result = Concurrency.cast_the_die()
        assert result >= 1 and result <= 6
      end
    end

    test "cast_dice_parallel で複数のサイコロを振る" do
      result = Concurrency.cast_dice_parallel(3)
      assert result >= 3 and result <= 18
    end

    test "cast_dice_and_store で結果を保存" do
      {:ok, agent} = Agent.start_link(fn -> [] end)
      Concurrency.cast_dice_and_store(5, agent)
      results = Agent.get(agent, & &1)

      assert length(results) == 5
      assert Enum.all?(results, fn x -> x >= 1 and x <= 6 end)

      Agent.stop(agent)
    end
  end

  describe "バックグラウンドプロセス" do
    test "start_background でプロセスを起動" do
      {:ok, pid} = Concurrency.start_background(fn -> :ok end)
      assert is_pid(pid)
    end

    test "cancel でプロセスを停止" do
      {:ok, pid} = Concurrency.start_background(fn -> Process.sleep(:infinity) end)
      assert Process.alive?(pid)

      Concurrency.cancel(pid)
      Process.sleep(10)
      refute Process.alive?(pid)
    end

    test "start_repeating で繰り返し実行" do
      {:ok, agent} = Agent.start_link(fn -> 0 end)

      {:ok, pid} =
        Concurrency.start_repeating(
          fn -> Agent.update(agent, &(&1 + 1)) end,
          10
        )

      Process.sleep(50)
      count = Agent.get(agent, & &1)
      assert count > 0

      Concurrency.cancel(pid)
      Agent.stop(agent)
    end
  end

  describe "ProcessingCheckIns" do
    test "チェックイン処理を開始・停止" do
      cities = [
        %City{name: "Tokyo"},
        %City{name: "Sydney"},
        %City{name: "Dublin"}
      ]

      {:ok, processing} = Concurrency.start_processing(cities)
      Process.sleep(50)

      ranking = Concurrency.current_ranking(processing)
      assert is_list(ranking)

      Concurrency.stop_processing(processing)
    end

    test "空のリストでも動作" do
      {:ok, processing} = Concurrency.start_processing([])
      assert Concurrency.current_ranking(processing) == []
      Concurrency.stop_processing(processing)
    end
  end

  describe "count_evens/1" do
    test "偶数をカウント" do
      ios = Enum.map(1..10, fn n -> fn -> n end end)
      count = Concurrency.count_evens(ios)
      assert count == 5
    end

    test "空リスト" do
      assert Concurrency.count_evens([]) == 0
    end
  end

  describe "collect_for/3" do
    test "指定時間内にデータを収集" do
      results = Concurrency.collect_for(100, 10, fn -> :rand.uniform(100) end)
      assert is_list(results)
      assert length(results) > 0
    end
  end

  describe "apply_updates/1" do
    test "更新を並行して適用" do
      updates = [
        %Update{key: "a", value: 1},
        %Update{key: "b", value: 2},
        %Update{key: "c", value: 3}
      ]

      result = Concurrency.apply_updates(updates)
      assert Map.get(result, "a") == 1
      assert Map.get(result, "b") == 2
      assert Map.get(result, "c") == 3
    end
  end

  describe "demo_parallel_sleep/2" do
    test "並列スリープは合計時間より短い" do
      {time, _} = :timer.tc(fn -> Concurrency.demo_parallel_sleep(3, 50) end)
      # 50ms * 3 = 150ms 未満で完了するはず
      assert time < 150_000
    end
  end
end
